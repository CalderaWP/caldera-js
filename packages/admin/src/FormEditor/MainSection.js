import React, {Component} from 'react';

export const MainSection = ({title, className, children}) => (
	<div
		className={className}
	>
		<h2>{title}</h2>
		{children}
	</div>
);
