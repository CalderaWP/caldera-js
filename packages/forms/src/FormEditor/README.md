
## Horizontal Forms
This is a small wrapper around the `<CalderaForm>` component that is supplied an array of fields and each field is shown in its own row.

### Primary Use Case(s)
* A section of Caldera UI that has a horizontal form, form example:
    - Processor UI
    - Field Config UI
* The inspector controls section of a Gutenberg block.


### Props
### Generic Example
```jsx
import {HorizontalForm} from '@calderawp/components';
<HorizontalForm
    fields={[]}
    initialValues={{}}
    onClose={onClose}
    onChange={onChange}
    instanceId={'test-2'}
/>
```

### Components
There are components that wrap `<HorizontalForm />`. They generally add controls for the UI state -- open, close, minimize, expand.

#### `<Processor />`
This is the settings -- to be used in the Caldera Forms editor or similar situation for describe one process of a form.

```js
import {Processor} from '@calderawp/components';

const processor = {
    id: 'p1',
    type: 'mailchimp',
    label: 'Segment One',
    fields: [
    	
    ],
    conditionals: [
    	
    ]
};

<Processor
    fields={processor.fields}
    initialValues={{}}
    onClose={onClose} //called when processor close button is clicked
    onChange={onChange} //called when any field is changed
    initialActiveTab={'conditionals'}//defaults to settings
    instanceId={processor.id} //shou
/>
```
