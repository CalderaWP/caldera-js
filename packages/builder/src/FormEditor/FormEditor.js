import React, {Component, Fragment, createElement} from 'react';
import PropTypes from 'prop-types';
import {Processors, processorsCollectionPropType} from './Processors/Processors';
import {TabPanel} from '@wordpress/components';
import {Row, Column,fieldAreaFactory} from "@calderajs/components";
import {processorTypesPropType} from './propTypes';
import {MainSection} from './MainSection';
import {defaultProcessorTypes} from './Processors/processorTypes/defaultProcessorTypes';
import {FormEntryViewer} from "..";

export class FormEditor extends Component {
	state = {
		activeTab: 'processors',
		newProcessorType: '', //the next processor to be created will use this type
	};

	onSetTab = activeTab => {
		this.setState({activeTab});
	};

	updateProcessors = processors => {
		const {updateForm, form} = this.props;
		updateForm({
			...form,
			processors
		});
	};

	getFormProcessors = () => {
		const {form} = this.props;
		return form.processors ? form.processors : [];
	};

	getFormFields = () => {
		const {form} = this.props;
		return form.fields ? form.fields : [];
	};

	tabs = [
		{
			name: 'editor',
			title: 'Layout',
			className: 'caldera-forms-editor-layout-tab-btn',
			classNameForComponent: 'caldera-forms-editor-layout'
		},
		{
			name: 'entries',
			title: 'Entries',
			className: 'caldera-forms-editor-entries-tab-btn',
			classNameForComponent: 'caldera-forms-editor-entries'
		},
		{
			name: 'processors',
			title: 'Processors',
			className: 'caldera-forms-editor-processors-tab-btn',
			classNameForComponent: 'caldera-forms-editor-processors'

		},
		{
			name: 'mailer',
			title: 'Mail',
			className: 'caldera-forms-editor-mailer-tab-btn',
			classNameForComponent: 'caldera-forms-editor-mailer'
		},
		{
			name: 'settings',
			title: 'Settings',
			className: 'caldera-forms-editor-settings-tab-btn',
			classNameForComponent: 'caldera-forms-editor-settings'
		},
	];

	getTabs(){
		const {hideTabs} = this.props;
		if( ! hideTabs.length ){
			return this.tabs;
		}

		return  this.tabs.filter( tab =>!hideTabs.includes(tab.name) );
	}


	setNewProcessorType = (newProcessorType) => this.setState({newProcessorType});

	render() {
		const {form, processorTypes, updateForm,entries} = this.props;
		const theProcessorTypes = [
			...processorTypes,
			...defaultProcessorTypes
		];
		return (
			<div>
				<Row>
					<Column>
						<p>{form.name}</p>
					</Column>
				</Row>
				<TabPanel
					className="caldera-processor"
					activeClass="active-tab"
					onSelect={this.onSetTab}
					initialTabName={'processors'}
					tabs={this.getTabs()}
				>
					{tab => {
						const {name, classNameForComponent, title} = tab;
						if ('processors' === name) {
							return (
								<MainSection
									className={classNameForComponent}
									title={title}
								>
									<Processors
										processorTypes={theProcessorTypes}
										processors={this.getFormProcessors()}
										form={form}
										formFields={this.getFormFields()}
										updateProcessors={this.updateProcessors}
									/>
								</MainSection>
							);
						}
						if ('settings' === name) {
							const nameField = {
								fieldType: 'text',
								value: form.name,
								label: 'Form Name',
								fieldId: 'formName',
								required: true,
							};
							return (
								<MainSection
									className={classNameForComponent}
									title={title}
								>
									{fieldAreaFactory(
										nameField,
										(name) => {
											updateForm({
												...form,
												name
											})
										}
									)}
								</MainSection>
							)
						}
						if ('entries' === name) {
							let entryViewerProps ={
								form,
								noItemsMessage: `No Entries Found For ${form.id}`,
								entries: {}
							};

							return (
								<MainSection
									className={classNameForComponent}
									title={title}
								>
									<FormEntryViewer {...entryViewerProps} />
								</MainSection>
							)
						}
						return (
							<MainSection
								className={classNameForComponent}
								title={title}
							>
								{tab.title}
							</MainSection>
						);
					}}
				</TabPanel>
			</div>
		);
	}
}


FormEditor.propTypes = {
	processorsTypes: processorTypesPropType,
	updateForm: PropTypes.func,
	form: PropTypes.shape({
		id: PropTypes.string,
		name: PropTypes.string,
		fields: PropTypes.array,
		conditionals: PropTypes.array,
		settings: PropTypes.array,
		processors: processorsCollectionPropType,
	}),
	hideTabs: PropTypes.array
};

FormEditor.defaultProps = {
	processorTypes: [],
	updateForm: () => {},
	hideTabs: []
};

