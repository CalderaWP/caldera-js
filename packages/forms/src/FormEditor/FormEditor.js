import React, {Component, Fragment, createElement} from 'react';
import PropTypes from 'prop-types';
import {Processors, processorsCollectionPropType} from './Processors/Processors';
import {TabPanel} from '@wordpress/components';
import {Row, Column,fieldAreaFactory} from "@calderawp/components";
import {processorTypesPropType} from './propTypes';
import {MainSection} from './MainSection';
import {defaultProcessorTypes} from './Processors/processorTypes/defaultProcessorTypes';

export class FormEditor extends Component {
	state = {
		activeTab: 'processors',
		newProcessorType: '', //the next processor to be created will use this type
	};

	onSetTab = activeTab => {
		this.setState({activeTab});
	};

	updateProcessors = processors => {
		const {updateForm, form} = this.props;
		updateForm({
			...form,
			processors
		});
	};

	getFormProcessors = () => {
		const {form} = this.props;
		return form.processors ? form.processors : [];
	};

	getFormFields = () => {
		const {form} = this.props;
		return form.fields ? form.fields : [];
	};

	tabs = [
		{
			name: 'editor',
			title: 'Layout',
			className: 'caldera-forms-editor-layout-tab-btn',
			classNameForComponent: 'caldera-forms-editor-layout'
		},
		{
			name: 'processors',
			title: 'Processors',
			className: 'caldera-forms-editor-processors-tab-btn',
			classNameForComponent: 'caldera-forms-editor-processors'

		},
		{
			name: 'Mailer',
			title: 'mailer',
			className: 'caldera-forms-editor-mailer-tab-btn',
			classNameForComponent: 'caldera-forms-editor-mailer'
		},
		{
			name: 'settings',
			title: 'Settings',
			className: 'caldera-forms-editor-settings-tab-btn',
			classNameForComponent: 'caldera-forms-editor-settings'

		},
	];


	setNewProcessorType = (newProcessorType) => this.setState({newProcessorType});

	render() {
		const {form, processorTypes, updateForm} = this.props;
		const theProcessorTypes = [
			...processorTypes,
			...defaultProcessorTypes
		];
		return (
			<div>
				<Row>
					<Column>
						<p>{form.name}</p>
					</Column>
				</Row>
				<TabPanel
					className="caldera-processor"
					activeClass="active-tab"
					onSelect={this.onSetTab}
					initialTabName={'processors'}
					tabs={this.tabs}
				>
					{tab => {
						const {name, classNameForComponent, title} = tab;
						if ('processors' === name) {
							return (
								<MainSection
									className={classNameForComponent}
									title={title}
								>
									<Processors
										processorTypes={theProcessorTypes}
										processors={this.getFormProcessors()}
										form={form}
										formFields={this.getFormFields()}
										updateProcessors={this.updateProcessors}
									/>
								</MainSection>
							);
						}
						if ('settings' === name) {
							const nameField = {
								fieldType: 'text',
								value: form.name,
								label: 'Form Name',
								fieldId: 'formName',
								required: true,
							};
							return (
								<MainSection
									className={classNameForComponent}
									title={title}
								>
									{fieldAreaFactory(
										nameField,
										(name) => {
											updateForm({
												...form,
												name
											})
										}
									)}
								</MainSection>
							)
						}
						return (
							<MainSection
								className={classNameForComponent}
								title={title}
							>
								{tab.title}
							</MainSection>
						);
					}}
				</TabPanel>
			</div>
		);
	}
}


FormEditor.propTypes = {
	processorsTypes: processorTypesPropType,
	updateForm: PropTypes.func,
	form: PropTypes.shape({
		id: PropTypes.string,
		name: PropTypes.string,
		fields: PropTypes.array,
		conditionals: PropTypes.array,
		settings: PropTypes.array,
		processors: processorsCollectionPropType,
	}),
};

FormEditor.defaultProps = {

}

