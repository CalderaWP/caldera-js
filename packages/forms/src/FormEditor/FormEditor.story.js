import React, {Component} from 'react';
import { storiesOf } from '@storybook/react';
import { Processor } from './Processors/Processor';
import { Processors } from './Processors/Processors';
import {processorsCollection} from './Processors/processors.fixtures';

import { FormEditor } from './FormEditor';


import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField
} from "@calderawp/components";
storiesOf('FormEditor', module).add('The processor', () => (
	<Processor
		fields={[
			checkboxFieldset,
			selectField,
			checkboxField,
			numberField,
			textField,
			emailField,
			radioField
		]}
		type={'mailchimp'}
		label={'Main Segment'}
		onChange={values => console.log(values)}
		onClose={values => console.log(values)}
		onRemove={values => console.log(values)}
	/>
));

const processorTypes = [
	{
		type: 'apiRequest'
	},
	{
		type: 'redirect'
	}
];

storiesOf('FormEditor', module).add('The processors list', () => (
	<Processors
		processorTypes={processorTypes}
		processors={processorsCollection}
		onChange={values => console.log(values)}
		onClose={values => console.log(values)}
		onRemove={values => console.log(values)}
		formFields={[]}
		form={ {id: 'Name'}}
	/>
));

const form = {
	id: 'a form',
	processors: processorsCollection,
};

class MockFormApp extends Component {
	state = {
		form: {
			id: 'a-form',
			name: 'Form Name',
			processors: processorsCollection,
			fields: [
				checkboxField,
				radioField,
				textField
			]
		}
	};

	updateForm = (form) => this.setState({form});

	render(){
		const {form} = this.state;
		return(
			<FormEditor
				processorTypes={processorTypes}
				updateForm={this.updateForm}
				form={ form }
			/>
		)
	}
}

storiesOf('FormEditor', module).add('The form editor', () => (
	<MockFormApp/>
));


