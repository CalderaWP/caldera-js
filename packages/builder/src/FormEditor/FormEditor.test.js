import React from 'react';
import { render } from 'react-testing-library';
import { FormEditor } from './FormEditor';
import { processorsCollection } from './Processors/processors.fixtures';
const form = {
	id: 'a-form',
	name: 'Form Name',
	processors: processorsCollection,
};

const processorTypes = [
	{
		type: 'apiRequest',
	},
	{
		type: 'redirect',
	},
];

let updateForm = jest.fn();

describe('FormEditor', () => {
	beforeEach(() => {
		updateForm = jest.fn();
	});

	it('should match snapshot', () => {
		const component = render(
			<FormEditor
				processorTypes={processorTypes}
				updateForm={updateForm}
				form={form}
			/>
		);
		expect(component).toMatchSnapshot();
	});
});
