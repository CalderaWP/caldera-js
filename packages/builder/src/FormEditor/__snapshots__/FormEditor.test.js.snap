// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormEditor should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      aria-atomic="true"
      aria-live="polite"
      aria-relevant="additions text"
      class="a11y-speak-region"
      id="a11y-speak-polite"
      style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
    />
    <div
      aria-atomic="true"
      aria-live="assertive"
      aria-relevant="additions text"
      class="a11y-speak-region"
      id="a11y-speak-assertive"
      style="position: absolute;margin: -1px;padding: 0;height: 1px;width: 1px;overflow: hidden;clip: rect(1px, 1px, 1px, 1px);-webkit-clip-path: inset(50%);clip-path: inset(50%);border: 0;word-wrap: normal !important;"
    />
    <div>
      <div>
        <div
          class="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
        >
          <div
            class="caldera-column width-1 sc-bdVaJa iHZvIS"
          >
            <p>
              Form Name
            </p>
          </div>
        </div>
        <div
          class="caldera-processor"
        >
          <div
            aria-orientation="horizontal"
            class="components-tab-panel__tabs"
            role="tablist"
          >
            <button
              aria-controls="0-editor-view"
              aria-selected="false"
              class="components-button caldera-forms-editor-layout-tab-btn "
              id="0-editor"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Layout
            </button>
            <button
              aria-controls="0-entries-view"
              aria-selected="false"
              class="components-button caldera-forms-editor-entries-tab-btn "
              id="0-entries"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Entries
            </button>
            <button
              aria-controls="0-processors-view"
              aria-selected="true"
              class="components-button caldera-forms-editor-processors-tab-btn active-tab"
              id="0-processors"
              role="tab"
              type="button"
            >
              Processors
            </button>
            <button
              aria-controls="0-mailer-view"
              aria-selected="false"
              class="components-button caldera-forms-editor-mailer-tab-btn "
              id="0-mailer"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Mail
            </button>
            <button
              aria-controls="0-settings-view"
              aria-selected="false"
              class="components-button caldera-forms-editor-settings-tab-btn "
              id="0-settings"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Settings
            </button>
          </div>
          <div
            aria-labelledby="0-processors"
            class="components-tab-panel__tab-content"
            id="0-processors-view"
            role="tabpanel"
            tabindex="0"
          >
            <div
              class="caldera-forms-editor-processors"
            >
              <h2>
                Processors
              </h2>
              <div>
                <div>
                  <div
                    class="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
                  >
                    <button
                      class="caldera-forms-choose-processor caldera-forms-choose-processor-season7Episode2"
                    >
                      Test sending form data to test API
                    </button>
                  </div>
                  <div
                    class="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
                  >
                    <button
                      class="caldera-forms-choose-processor caldera-forms-choose-processor-p2"
                    >
                      Main MailChimp
                    </button>
                  </div>
                </div>
                <div>
                  <div
                    class="caldera-field-group caldera-field-area-select is-required form-group"
                  >
                    <label
                      class="form-label"
                      for="newProcessorType"
                    >
                      Processor Type
                       
                      <span>
                        *
                      </span>
                    </label>
                    <select
                      class="caldera-field caldera-field-select form-control"
                      id="newProcessorType"
                      required=""
                    >
                      <option>
                        --
                      </option>
                      <option>
                        --
                      </option>
                      <option
                        value="apiRequest"
                      >
                        apiRequest
                      </option>
                      <option
                        value="redirect"
                      >
                        redirect
                      </option>
                      <option
                        value="autoResponder"
                      >
                        autoResponder
                      </option>
                      <option
                        value="apiRequest"
                      >
                        apiRequest
                      </option>
                    </select>
                    
                  </div>
                  <button
                    disabled=""
                  >
                    Add 
                    
                     Processor
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
      >
        <div
          class="caldera-column width-1 sc-bdVaJa iHZvIS"
        >
          <p>
            Form Name
          </p>
        </div>
      </div>
      <div
        class="caldera-processor"
      >
        <div
          aria-orientation="horizontal"
          class="components-tab-panel__tabs"
          role="tablist"
        >
          <button
            aria-controls="0-editor-view"
            aria-selected="false"
            class="components-button caldera-forms-editor-layout-tab-btn "
            id="0-editor"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Layout
          </button>
          <button
            aria-controls="0-entries-view"
            aria-selected="false"
            class="components-button caldera-forms-editor-entries-tab-btn "
            id="0-entries"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Entries
          </button>
          <button
            aria-controls="0-processors-view"
            aria-selected="true"
            class="components-button caldera-forms-editor-processors-tab-btn active-tab"
            id="0-processors"
            role="tab"
            type="button"
          >
            Processors
          </button>
          <button
            aria-controls="0-mailer-view"
            aria-selected="false"
            class="components-button caldera-forms-editor-mailer-tab-btn "
            id="0-mailer"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Mail
          </button>
          <button
            aria-controls="0-settings-view"
            aria-selected="false"
            class="components-button caldera-forms-editor-settings-tab-btn "
            id="0-settings"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Settings
          </button>
        </div>
        <div
          aria-labelledby="0-processors"
          class="components-tab-panel__tab-content"
          id="0-processors-view"
          role="tabpanel"
          tabindex="0"
        >
          <div
            class="caldera-forms-editor-processors"
          >
            <h2>
              Processors
            </h2>
            <div>
              <div>
                <div
                  class="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
                >
                  <button
                    class="caldera-forms-choose-processor caldera-forms-choose-processor-season7Episode2"
                  >
                    Test sending form data to test API
                  </button>
                </div>
                <div
                  class="sc-bdVaJa caldera-row sc-bwzfXH pjOvZ"
                >
                  <button
                    class="caldera-forms-choose-processor caldera-forms-choose-processor-p2"
                  >
                    Main MailChimp
                  </button>
                </div>
              </div>
              <div>
                <div
                  class="caldera-field-group caldera-field-area-select is-required form-group"
                >
                  <label
                    class="form-label"
                    for="newProcessorType"
                  >
                    Processor Type
                     
                    <span>
                      *
                    </span>
                  </label>
                  <select
                    class="caldera-field caldera-field-select form-control"
                    id="newProcessorType"
                    required=""
                  >
                    <option>
                      --
                    </option>
                    <option>
                      --
                    </option>
                    <option
                      value="apiRequest"
                    >
                      apiRequest
                    </option>
                    <option
                      value="redirect"
                    >
                      redirect
                    </option>
                    <option
                      value="autoResponder"
                    >
                      autoResponder
                    </option>
                    <option
                      value="apiRequest"
                    >
                      apiRequest
                    </option>
                  </select>
                  
                </div>
                <button
                  disabled=""
                >
                  Add 
                  
                   Processor
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
